### ISSUE TITLE
Bitmessage v0.6.1 Windows client (32bit and 64bit) crash at startup in module nvopencl.dll

### ISSUE THREAD

**Issue by mothe-at** [Bitmessage/PyBitmessage#906]
I am running Windows 10 on a notbook with a built-in NVIDIA Quadro K1000M.
OS and all drivers are all up to date.
When I try to start the Bitmessage v0.6.1 client it never appears on screen. In the Application Event Log there is this crash-report:

```
Source: Application Error
Event ID: 1000

Faulting application name: Bitmessage-0.6.1.exe, version: 0.0.0.0, time stamp: 0x57b1d704
Faulting module name: nvopencl.dll, version: 6.14.13.6909, time stamp: 0x579f3ea8
Exception code: 0xc0000005
Fault offset: 0x00016a65
Faulting process id: 0x17f0
Faulting application start time: 0x01d231c071807e2a
Faulting application path: D:\TEMP\Bitmessage-0.6.1.exe
Faulting module path: C:\WINDOWS\System32\nvopencl.dll
Report Id: 37467aa8-96cc-4729-8884-ec7d27474fbb
Faulting package full name: 
Faulting package-relative application ID:
```

It happens with the 32bit and the 64bit client.
The Bitmessage v0.6.0 client works fine, no crash.
Have played around with several driver versions, compatibility modes, etc. No way.

Any idea or any log I can share?

**Comment by PeterSurda** [comment]
If the notebook has a hybrid GPU setup, try to tell the nvidia control panel to launch bitmessage on the onboard (presumably intel) card instead of the nvidia card.

**Comment by mothe-at** [comment]
Yes it has a hybrid GPU, but running it on the integrated Intel chip causes the same crash (either with "right-click > Run with graphics processor > Integrated graphics (default)" or with setting BM to run on the integrated chip in the NVIDIA control panel).
Thank you!

**Comment by PeterSurda** [comment]
What about other OpenCL apps, like [CompuBench CL](https://compubench.com/windows-download/compubench/)?

**Comment by mothe-at** [comment]
CompuBench CL works, all tests successful.

**Comment by PeterSurda** [comment]
I tried to reproduce it on a laptop with Win 10 and Intel/740M hybrid and I can't. Bitmessage can't initalise OpenCL but it doesn't crash, it just keeps it off. Did you select the Nvidia card for CompuBench?

The binary for 0.6.0 that used to be on the bitmessage.org page doesn't have a working OpenCL so that explains why it works.

**Comment by PeterSurda** [comment]
I remember one Windows developer reported to me that there are problems if you have multiple GPUs with different platforms/drivers, but again I was never able to reproduce it. It looks like you might be experiencing this. Would you be willing to test a custom built binary that I'll send you?

**Comment by mothe-at** [comment]
Sure I am willing to help. Just send me the Link to BM-2cXtYfFSf65wkgkY6ZrsCEvib9dExWLZpr

**Comment by PeterSurda** [comment]
Message sent.

**Comment by PeterSurda** [comment]
So it looks like the cross-platform OpenCL issue it was indeed. However it behaves inconsistently, on some systems, it simply ends up with OpenCL being disabled, on others it causes a crash. I'll add a platform selector, but it either needs to be multiple-choice, or a smart combo, because in some cases you may want to use multiple GPUs from the same vendor, in other cases you may not.

**Comment by mothe-at** [comment]
Thank you, Peter, for the quick fix! Meanwhile I can use the custom build, works fine!

**Comment by PeterSurda** [comment]
Should be fixed in e9a3ef465ca68dd182b54437f061944a9d106786 but it's difficult to test for me at the moment. I'll do testing/debugging before releasing 0.6.2.

**Comment by PeterSurda** [comment]
Please test the development snapshots (the latest version):

https://bitmessage.org/download/snapshots/win64/
https://bitmessage.org/download/snapshots/win32/

They allow you to choose the OpenCL vendor in the "Maximum acceptable difficulty". I got positive feedback but I haven't tested it myself yet.

**Comment by PeterSurda** [comment]
I tested it on my Intel/Nvidia laptop and it works.

**Comment by mothe-at** [comment]
Hi Peter!
Tested Bitmessagedev_x64_20170301.exe and Bitmessagedev_x86_20170301.exe, both worked well!
Thank you!

**Comment by PeterSurda** [comment]
You're welcome. Also you can now upgrade to 0.6.2 which has 1 or 2 more fixes compared to the 20170301 snapshot.
